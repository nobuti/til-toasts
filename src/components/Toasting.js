/** @jsx jsx */
import { useContext, useState } from 'react'
import { jsx, css } from '@emotion/core'
import { useTransition, animated } from 'react-spring'

import ToastContext from './toastContext'
import Toast from './Toast'

const Toasting = () => {
  const { toasts } = useContext(ToastContext)
  const [refMap] = useState(() => new Map())

  const transitions = useTransition(toasts, (toast) => toast.toast, {
    enter: (item) => async (next) => {
      await next({
        opacity: 1,
        height: refMap.get(item.toast).clientHeight + 16,
      })
    },
    leave: (item) => async (next) => {
      refMap.delete(item.toast)
      await next({ opacity: 0, height: 0 })
    },
    from: { opacity: 0, height: 0 },
  })

  return (
    <div
      css={css`
        position: fixed;
        bottom: 16px;
        right: 16px;
      `}
    >
      <ul>
        {transitions.map(({ item, props, key }) => (
          <animated.li key={key} style={props}>
            <div ref={(ref) => ref && refMap.set(item.toast, ref)}>
              <Toast key={item.toast} {...item} />
            </div>
          </animated.li>
        ))}
      </ul>
    </div>
  )
}

export default Toasting
