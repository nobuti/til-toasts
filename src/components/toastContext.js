import React, { createContext } from 'react'

import useToast from '../hooks/useToast'
const ToastContext = createContext(null)

export const ToastProvider = ({ children }) => {
  const toast = useToast()
  return <ToastContext.Provider value={toast}>{children}</ToastContext.Provider>
}

export default ToastContext
