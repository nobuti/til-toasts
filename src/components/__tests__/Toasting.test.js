import React from 'react'
import { render } from '@testing-library/react'

import Toasting from '../Toasting'
import * as ToastContext from '../toastContext'

describe('Toasting', () => {
  jest.spyOn(ToastContext, 'useToastContext').mockImplementation(() => ({
    toasts: [
      {
        toast: 1,
        color: '#2548f5',
        title: 'Wadus',
        message: 'Wadus wadus!',
      },
      {
        toast: 2,
        color: '#fabada',
        title: 'Bogus',
        message: 'Bogus wadus!',
      },
    ],
  }))

  it('should render properly', () => {
    const { container } = render(<Toasting />)
    expect(container).toMatchSnapshot()
  })
})
