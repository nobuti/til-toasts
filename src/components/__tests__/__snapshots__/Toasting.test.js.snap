// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toasting should render properly 1`] = `
<div>
  <div
    class="css-3tqauy-Toasting"
  >
    <ul>
      <li
        style="opacity: 0; height: 0px;"
      >
        <div>
          <div
            class="css-m76tdx-Toast"
          >
            <h2
              class="css-1v9h6x6-Toast"
            >
              Wadus
            </h2>
            <p
              class="css-1jfni4t-Toast"
            >
              Wadus wadus!
            </p>
            <button
              class="css-1nw5hej-Toast"
            >
              <svg>
                x.svg
              </svg>
            </button>
          </div>
        </div>
      </li>
      <li
        style="opacity: 0; height: 0px;"
      >
        <div>
          <div
            class="css-10tjqg4-Toast"
          >
            <h2
              class="css-1v9h6x6-Toast"
            >
              Bogus
            </h2>
            <p
              class="css-1jfni4t-Toast"
            >
              Bogus wadus!
            </p>
            <button
              class="css-1nw5hej-Toast"
            >
              <svg>
                x.svg
              </svg>
            </button>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
